### ISSUE TITLE
Add HavlockV/CoC support

### ISSUE THREAD

**Issue by lozalojo** [League-of-Foundry-Developers/fvtt-module-polyglot#95]
If possible, add support for HavlockV/CoC system.

https://github.com/HavlockV/CoC7-FoundryVTT

**Comment by mclemente** [comment]
How does languages work on CoC7? Are they skills?

**Comment by lozalojo** [comment]
> How does languages work on CoC7? Are they skills?

Yes, Own Language + Other Language/Specializations: English, French, Latin, Greek

Specializations are a feature of the system

**Comment by mclemente** [comment]
Is there a field for Own Language? On the foundry system, Other Languages are under the "Language (Any)" skill, right?

**Comment by lozalojo** [comment]
> Is there a field for Own Language? On the foundry system, Other Languages are under the "Language (Any)" skill, right?

Yes, there are two different skills:

Own Language
Other language/specialization

Since skills are fully customizable, you can create Own Language as another specialization skills, setting the spectialization value yo the actual own language of the pc, in case it is needed for Polyglot to work.

Alson skill names will differ between localization, which I asume you've already taken into consideration.

**Comment by mclemente** [comment]
Could you try this version: https://github.com/League-of-Foundry-Developers/fvtt-module-polyglot/archive/refs/heads/nightly.zip 
It's the beta build, just overwrite polyglot folder with its contents. It should work for languages on character sheets (e.g. `Language (Test)` becomes `test`), but won't show languages on journals or when the Keeper has no character selected since the system doesn't have the languages set up, unless you set them up yourself as custom languages.
I think custom languages written by the Keeper can be read by players as long as it uses the same case (e.g. if you set up `test` as a custom language, it might work, but `Test` definitely won't).

**Comment by lozalojo** [comment]
Not sure if I did setup correctly: I've created custom languages English, Spanish and French:

GM view:
![](https://i.ibb.co/FJT2Xm1/screenshot-118.jpg)

Ive writen 3 messages in different languages.
Players view: this one only understand Spanish, but can read all messages

![](https://i.ibb.co/nQtDTdK/screenshot-117.jpg)

**Comment by lozalojo** [comment]
This is the GM selecting a token, and speaking in French, the player side screenshot, it is supposed not to understand French:

![](https://i.ibb.co/27wL2Y3/screenshot-119.jpg)

**Comment by mclemente** [comment]
Try the setting the custom languages in lowercase instead (i.e. english, french, spanish).

**Comment by lozalojo** [comment]
> Try the setting the custom languages in lowercase instead (i.e. english, french, spanish).

Same, that's the first thing I tried, always translate the text to the client, the clients owns a single actor with Language (Spanish) but it is able to read all languages.

**Comment by mclemente** [comment]
https://github.com/League-of-Foundry-Developers/fvtt-module-polyglot/issues/95#issuecomment-803527436
Were you using the `Scramble on OOC chat messages` setting? Otherwise, non-character messages will be "plain text" instead of being written on a language.

**Comment by lozalojo** [comment]
> [#95 (comment)](https://github.com/League-of-Foundry-Developers/fvtt-module-polyglot/issues/95#issuecomment-803527436)
> Were you using the `Scramble on OOC chat messages` setting? Otherwise, non-character messages will be "plain text" instead of being written on a language.

Was turned off, I turned on and same, the client can see the translation card independtly of the language used.
I also used /ic and /ooc messages, same for both

**Comment by mclemente** [comment]
I've just updated it, try replacing the file again https://github.com/League-of-Foundry-Developers/fvtt-module-polyglot/archive/refs/heads/nightly.zip

**Comment by lozalojo** [comment]
> I've just updated it, try replacing the file again https://github.com/League-of-Foundry-Developers/fvtt-module-polyglot/archive/refs/heads/nightly.zip

No luck, Harvey is speaking in three different languages and Player 1 understand all of them:

![](https://i.ibb.co/VTwsCC5/screenshot-93.jpg)

Does it work for you? Have you tried with CoC7th? Maybe I've not configured skills correctly.

**Comment by mclemente** [comment]
I see. I've used the Language (Any/Own) rather than Language (Other) and they function differently.
Is there any different between then on the game's mechanics? I'll try setting something for Language (Other).

Just to be sure what I've done works for you: try creating a `Language (English)` skill based on Language (Own/Any) and move it to your character,

**Comment by mclemente** [comment]
> https://github.com/League-of-Foundry-Developers/fvtt-module-polyglot/archive/refs/heads/nightly.zip

Grab that file again, it should work for both types now.

**Comment by lozalojo** [comment]
> I see. I've used the Language (Any/Own) rather than Language (Other) and they function differently.
> Is there any different between then on the game's mechanics? I'll try setting something for Language (Other).

No differences on game mechanics, they are different only because they have different base values.

> Just to be sure what I've done works for you: try creating a `Language (English)` skill based on Language (Own/Any) and move it to your character,
> 

![](https://i.ibb.co/VW0kRDh/screenshot-122.jpg)

> EDIT: OMG I just noticed I've sent you the wrong link that second time, I've changed the branch's name to CoC7 and didn't change the link. You've got a beta version that didn't even had support for CoC7.

No prob, I was already using CoC branch, I noticed that there were no commits on nightly branch and saw CoC.

Using "version": "1.5.6.10", shouldnt be ```"POLYGLOT.COC7.LanguageSkills": "Language"``` instead of ```"POLYGLOT.COC7.LanguageSkills": "Language"```

Same, Harvey speaking (as GM selecting the actor), Michael reading (as a client owning the actor) with all above skills possibilities, and he keeps on being able to read all messages:

![](https://i.ibb.co/MDYhZGL/screenshot-121.jpg)

Do you mind confirming that it really works on your CoC tests? I'd like to know if it is a problem of my setup, not the module.

**Comment by mclemente** [comment]
> Using "version": "1.5.6.10", shouldnt be `"POLYGLOT.COC7.LanguageSkills": "Language (Other)"` instead of `"POLYGLOT.COC7.LanguageSkills": "Language"`

Are you using Babele or any translatator? CoC7 uses `Language` for the specialization's name, not `Language (Other)`.

![image](https://user-images.githubusercontent.com/5288872/112015142-4ea06380-8b0a-11eb-9f61-3c139ed7b18e.png)
![image](https://user-images.githubusercontent.com/5288872/112015340-7ee80200-8b0a-11eb-91c7-36472b0fc2de.png)

> Do you mind confirming that it really works on your CoC tests? I'd like to know if it is a problem of my setup, not the module.

No language:
![image](https://user-images.githubusercontent.com/5288872/112013589-e735e400-8b08-11eb-8e80-47744ab87d69.png)

English and Language (French).
![image](https://user-images.githubusercontent.com/5288872/112014077-54e21000-8b09-11eb-8ae7-1f630af9924a.png)

English and Spanish.
![image](https://user-images.githubusercontent.com/5288872/112014412-9e325f80-8b09-11eb-9ffb-f1f9954b9bdb.png)

**Comment by lozalojo** [comment]
Hey, I've found the bug, let me record the screen, it is easier than explaining it

**Comment by lozalojo** [comment]
User Player 3 owns an actor called Lovely.

![](https://i.ibb.co/fHN3DNn/screenshot-129.jpg)
![](https://i.ibb.co/C29QKjn/screenshot-132.jpg)

More info on Lovely's skills:

```"items":[{"_id":"iGVSojqjsdXzg3LE","name":"English","type":"skill","data":{"specialization":"Language","description":{"value":"","opposingDifficulty":"","pushedFaillureConsequences":"","chat":""},"base":"1","adjustments":{"personal":null,"occupation":null,"archetype":null,"experience":null},"value":null,"attributes":{},"properties":{"special":true,"push":true},"eras":{},"flags":{}},"sort":100001,"flags":{"core":{"sourceId":"Item.EM0QYPJ9Qr9sIpiv"}},"img":"icons/svg/mystery-man.svg","effects":[]},{"_id":"YuarS40kwwS8OkPd","name":"French","type":"skill","data":{"specialization":"Language","description":{"value":"","opposingDifficulty":"","pushedFaillureConsequences":"","chat":""},"base":"1","adjustments":{"personal":null,"occupation":null,"archetype":null,"experience":null},"value":null,"attributes":{},"properties":{"special":true,"push":true},"eras":{},"flags":{}},"sort":100001,"flags":{"core":{"sourceId":"Item.FnUlbGvJMsKHlNEQ"}},"img":"icons/svg/mystery-man.svg","effects":[]}]```

This video is from the client side (Player 3 owning Lovely):

![](https://www.youtube.com/watch?v=ZSvrpy2jD-U)

More info on my setup:

![](https://i.ibb.co/3zw8Gvq/screenshot-128.jpg)
![](https://i.ibb.co/SB949dW/screenshot-130.jpg)
![](https://i.ibb.co/mJgcswL/screenshot-131.jpg)
![](https://i.ibb.co/ykJVcYC/screenshot-127.jpg)

**Comment by mclemente** [comment]
Are you sure Lovely's owner isn't set as Harvey's owner too?
I've created two character with different sets of languages and the only way I can reproduce what happens on the video is if I set the player as the owner of both characters.

**Comment by lozalojo** [comment]
> Are you sure Lovely's owner isn't set as Harvey's owner too?
> I've created two character with different sets of languages and the only way I can reproduce what happens on the video is if I set the player as the owner of both characters.

Yes, Player 3 has ownership over Harvey also. So, is it intentional? To be able to switch between actors which you have owner permissions, not only the one you currently own?

**Comment by mclemente** [comment]
It's the same mechanic as a GM. If they aren't supposed to see those, they should be Observers, not Owners.

**Comment by lozalojo** [comment]
> It's the same mechanic as a GM. If they aren't supposed to see those, they should be Observers, not Owners.

**Comment by mclemente** [comment]
Just updated nightly, thank you for testing everything.

**Comment by lozalojo** [comment]
Thank you, just for clarification, for localization (if I play in spanish), is it enough to change the key:

```"POLYGLOT.COC7.LanguageSkills": "Language"```

This affect the Language skill, but you mentioned that it also check the Language (Own), which is a different skill, isnt it required to specify a secondary skill for Own language?

Skill 1: Language (Own), specialization: English
Skill 2: Language, specialization: French, Spanish, Greek, etc...

**Comment by mclemente** [comment]
> 
> 
> Thank you, just for clarification, for localization (if I play in spanish), is it enough to change the key:
> 
> `"POLYGLOT.COC7.LanguageSkills": "Language"`

It should be the same used on the Babele translation, `Idioma`. And the skills should be named either Specialization: "Idioma" or "Idioma (Castellano)" works.
I'm not sure if that's how CoC's translation works, that's how WFRP and SWADE handle it.

> 
> This affect the Language skill, but you mentioned that it also check the Language (Own), which is a different skill, isnt it required to specify a secondary skill for Own language?
> 
> Skill 1: Language (Own), specialization: English
> Skill 2: Language, specialization: French, Spanish, Greek, etc...

The logic is basically
```
if (skill's name contains "Language")
 return text in between the parenthesis in skill's name.
else if (skill's specialization is "Language")
 return skill's name
```

So just have both the Language skills and Language specialization skill use the same name.

**Comment by lozalojo** [comment]
> > Thank you, just for clarification, for localization (if I play in spanish), is it enough to change the key:
> > `"POLYGLOT.COC7.LanguageSkills": "Language"`
> 
> It should be the same used on the Babele translation, `Idioma`. And the skills should be named either Specialization: "Idioma" or "Idioma (Castellano)".
> I'm not sure if that's how CoC's translation works, that's how WFRP and SWADE handle it, though.
> 
> > This affect the Language skill, but you mentioned that it also check the Language (Own), which is a different skill, isnt it required to specify a secondary skill for Own language?
> > Skill 1: Language (Own), specialization: English
> > Skill 2: Language, specialization: French, Spanish, Greek, etc...
> 
> The logic is basically
> 
> ```
> if (skill's name contains "Language")
>  return text in between the parenthesis in skill's name.
> else if (skill's specialization is "Language")
>  return skill's name
> ```
> 
> So just have both the Language skills and Language specialization skill use the same name.

Hi, I've just tested with my localization (Spanish) and I've got two issues:

```"POLYGLOT.COC7.LanguageSkills": "Lengua"```

In spanish the skills are "Lengua propia" and "Otras lenguas".

1. Only works when the skill is exactly "Lengua", does not work for me with "Lengua propia" neither "Otras lenguas".
2. I'd suggest not comparing casewise:
if (lowcase(skill's name) contains lowcase("POLYGLOT.COC7.LanguageSkills"))

In order to search for Lengua (Lengua propia - starting with uppercase) and lengua (Otras lenguas - starting with lowercase) .

And also to compare custom languages in the setup:

```"POLYGLOT.CustomLanguagesTitle"```

with lowercase versions, to be able to input Spanish/spanish, French/french in both, skills and configuration, and avoid typing problems and some headaches to users not being able to configure polyglot correctly.

Just a suggestion.

Thanks for your work!

**Comment by mclemente** [comment]
I'll break LanguageSkills into 4 (Own, Any, Other, Spec) so it's easier to set up.
The first 3 need to be the exact one used on your translation.
In case if doesn't find the terms, it looks for the spec.
Spec string should be a term people would use. It doesn't need to be the oficial term (it uses either this string, "Language", or the `CoC7.language` string).

**Comment by lozalojo** [comment]
Cant get it to work, current config:
![](https://i.ibb.co/cQ7qTR1/screenshot-94.jpg)
![](https://i.ibb.co/mGsSV3S/screenshot-95.jpg)
![](https://i.ibb.co/CwDz0Mw/screenshot-96.jpg)

**Comment by lozalojo** [comment]
More info on skill definition, this is for Greek, the name element does not include the text "Otras lenguas":

```{"_id":"UbhJjjbm5aMRXzNC","name":"Griego","permission":{"default":0,"DrIsbjsRd1cgvfpo":3},"type":"skill","data":{"specialization":"Otras lenguas","description":{```

**Comment by mclemente** [comment]
I still get confused some with which is a spec or not lol. Got it to work with both now, I guess.
Also, set `POLYGLOT.COC7.LanguageAny` to `Otras lenguas`.

**Comment by lozalojo** [comment]
I tested quickly this morning, didnt work. I will take a look again this afternoon.

> I still get confused with which is a spec or not lol. Try it again, got it to work with both now, I guess.
> Also, set `POLYGLOT.COC7.LanguageAny` to `Otras lenguas`.

Maybe I should have clarify this before.

In Cthulhu there are two skills related with languages:

Language (Own) - Lengua propia
Language (Other) - Otras lenguas

The system divides in two because both have different base values, Own base stat is based on an attribute (Education) and Other base stat is 1 (you have to learn other languages rather than your own). Thats why there are different skills for Own and Other.

Both skills are specialization skills, meaning that you have to specify which language you really know, Own only takes one specialization (your mother's language) and Others can have multiple languages.

All players have one own language and zero or more other languages.

CoC uses occupations, which is like the profession of the character. Occupation is a group of predefined skills linked to that occupation and where you can allocate some points in character creation.

Occupations in foundry are items themselves where you dragndrop skills to define the skills you can spend your points in.

When you dragndrop a specialization skills it will ask you for the specialization, for example, an Archaelogist have Language (Other), if I dragndrop that skill to the occupation archaelogist item, it will ask for an specialization. But I dont want to set an specialization at this point, because occupations are just the definition of professions, and every character might want to chose a different language to specialize. This is why CoC Foundry decided to add the specialization value "Any", meaning Generic:

Language (Own), specialization value: Any or Language (Own) (Any)
Language (Other), specialization value: Any or Language (Other) (Any)

So that when you dragndrop an Any named skill, the CoC system knows it is a generic specialization, and do not ask you for the specialization itself.

Why specialization value might be Any or Language (Own) (Any)? Technically should be any, but since at some places of the Foundry environment only the "name" is shown and not the "specialization", then if you see a skill called Any, you do not know if it is the generic version of Language (Other) or Science or Pilot or Survival (all are specialization skills), so that's why it is a good idea to repeat he specialization in the name: Language (Own) (Any) and Language (Other) (Any) so that you are sure that it is clear which specialization is the skill for.

The name "Any", then, it is just for creation mechanism, and shouldnt appear in a character sheet.

So, in summary, there are two skills for languages:

specialization: Language (Own), name: English, Spanish, French
specialization: Language (Other), name:  English, Spanish, French

If you have further questions, do not hesitate to ask.

**Comment by lozalojo** [comment]
Took a look at polyglot.js I do not know javascript, but

item.name.match <- does not contain Language (Own), just the languages names
item.data.specialization.match (not sure if I can address the structure of a js item this way) <- does contain Language (Own)

So, maybe something like this, sorry if this is stupid, I do not the structure of js objects and maybe it is a nonsense (also removed regexpr not really needed, in other languages would work, in js dont know):

```case "CoC7":
						for (let item of actor.data.items) {
							const match = 
								item.data.specialization.match( game.i18n.localize("POLYGLOT.COC7.LanguageOwn")+, 'i' )
								|| item.data.specialization.match( game.i18n.localize("POLYGLOT.COC7.LanguageOther")+, 'i' );
							// adding only the descriptive language name, not "Language (XYZ)"
							if (match)
								this.known_languages.add(item.name.trim().toLowerCase());
						}
						break;```

**Comment by mclemente** [comment]
The regex is supposed to remove the `Language ()` from the item's name.
If there's no match, it'll look for the item's specialization instead and then add its name.
Is it working correctly on your Spanish translation? If it is, it will probably work for any language.

**Comment by lozalojo** [comment]
> The regex is supposed to remove the `Language ()` from the item's name.
> If there's no match, it'll look for the item's specialization instead and then add its name.
> Is it working correctly on your Spanish translation? If it is, it will probably work for any language.

Not sure, but item.name only contains "French", "Spanish", it does not contain "Language (Other)"

According to my rationale:

item.name = "French"
item.data.specialization = "Language (Other)"

But I might be wrong

The item is like:

```"items":[{"_id":"iGVSojqjsdXzg3LE","name":"English","type":"skill","data":{"specialization":"Language (Other)","description":{"value":"","opposingDifficult```

**Comment by lozalojo** [comment]
> Is it working correctly on your Spanish translation? If it is, it will probably work for any language.

Nope, it does not work

**Comment by mclemente** [comment]
> Not sure, but item.name only contains "French", "Spanish", it does not contain "Language (Other)"

It's in case someone renames only the text within the parenthesis and plays with "Language (French)". Dunno why someone would do that, but it might happen.

> According to my rationale:
> 
> item.name = "French"
> item.data.specialization = "Language (Other)"

Are there more than one Language Specialization? I thought all were under Language. Anyway, I've added those to the list so it checks.

Also, I found out I used the wrong function to check for the specs. Try it now.

**Comment by lozalojo** [comment]
Ok, now it works, I think the "POLYGLOT.COC7.LanguageAny": "Language (Any)" could be removed, since it wont be used in the game, it is just for occupation's mechanics.

Also, I found an issue in the process, do I need to report it in the issue tracker?

If you do not own any player, but click on a token that you dont own, but has the language of the text, it is translated; I recorded a video:

https://www.youtube.com/watch?v=zPAi7ZBxoUY

Take a look in case you can replicate it.

Thanks!

**Comment by mclemente** [comment]
> Ok, now it works, I think the "POLYGLOT.COC7.LanguageAny": "Language (Any)" could be removed, since it wont be used in the game, it is just for occupation's mechanics.

Yes, I'll just leave it in case anything changes on CoC7 or people use the formatting. And the term might be different for each language, so who knows.

> Also, I found an issue in the process, do I need to report it in the issue tracker?

I managed to reproduce it, but only if the player has Owner permission on both tokens.

Anyway, I've accepted your update on the Spanish translation and merged nightly onto the main branch.

**Comment by lozalojo** [comment]
> I managed to reproduce it, but only if the player has Owner permission on both tokens.

Ok, as you see in the video the player didnt have owner permission on the tokens, but I've been making a lot of test in the pasts days, so maybe something got messed up in the world, dont know.
 
> Anyway, I've accepted your update on the Spanish translation and merged nightly onto the main branch.

Perfect, thanks
